### DISCUSSION TITLE
Value of field "db_file_id" does not exist on referenced table

### DISCUSSION THREAD

**Discussion by radynsade** [phalcon/phalcon#306] [Q&A] [answered]
Trying to use phalcon model relationships to save related models but get an error: "Value of field "db_file_id" does not exist on referenced table".

**Saving and processing services:**
[FilesService.php](https://gist.github.com/Radynsade/1a65ae8d7f941df71b1a0037d9d997c9)
[ImagesService.php](https://gist.github.com/Radynsade/fa1fff6960ce0447f6051e0d671a9c63)

**Base classes for models:**
[TraceableModel.php](https://gist.github.com/Radynsade/44ea45319491ff23f1e9040df5fc4eae)
[BaseModel.php](https://gist.github.com/Radynsade/91a4cd84705cfcfb99ac7de8722cfbe0)

**Models:**
[DbFile.php](https://gist.github.com/Radynsade/ddcc9ac682962144a1f8fc4273211d3b)
[Image.php](https://gist.github.com/Radynsade/0ff93fe3bb59090c0fada78821376561)

**Errors and logs:**
[SQL log](https://gist.github.com/Radynsade/725f1cbb2b5cd46430d2159cbe2b69cf)
[Error log](https://gist.github.com/Radynsade/6780ca8b5041caaae2c0c917b7bcac87)

**Comment by borisdelev** [comment]
Well.. little difficult... i cant understand whole picture, but... i see that "belongs to" in Image model:

```php
		$this->belongsTo(
			'db_file_id',
			DbFile::class,
			'id',
			[
				'alias' => 'dbFile',
				'foreignKey' => true,
				'reusable' => true
			]
		);
```

and in ImageService.php (see __upload__ function)... is not set. You have "__setDbFile__" function in image.php, but the db_file_id is not set there and is missing.

**Reply by radynsade** [reply]
But isn't a purpose of relationships that all id's handled automatically? As I read in manuals, in its examples they do not provide id's manually while saving with related models. Sorry if my message looks rude, just trying to figure out how it works. Anyway, I tried to set it manually and got the same error.

**Reply by borisdelev** [reply]
Well, depends. But no, you must provide it. Anyway... lets start from model description... did u have in table "images" field with name: db_file_id (probably yes, but i ask)? 
```protected int $dbFileId;``` must be public. If its protected, must be set with setter (for example). Try to change it.

I am trying to follow your logick... but its difficult in that case. And.... a question- why in model __Image.php__ setters and geters return static?

**Reply by radynsade** [reply]
Set accessor of $dbFileId to public, the same error. There are public setter and getter for this field. Maybe SQL definitions of the tables will help, I use PostgreSQL.
`

	CREATE SCHEMA "upload";

	CREATE TABLE IF NOT EXISTS "upload"."db_file" (
		"id" bigserial NOT NULL PRIMARY KEY,
		"link" varchar(256) NOT NULL,
		"mime_type" varchar(256) NOT NULL,
		"size" int NOT NULL,
		"created_at" timestamptz NOT NULL,
		"updated_at" timestamptz NOT NULL
	);

	CREATE TABLE IF NOT EXISTS "upload"."image" (
		"id" bigserial NOT NULL PRIMARY KEY,
		"is_vector" boolean NOT NULL,
		"db_file_id" bigint NOT NULL,
		CONSTRAINT fk_upload_image_upload_db_file
			FOREIGN KEY("db_file_id")
			REFERENCES "upload"."db_file"("id")
			ON DELETE CASCADE
	);

`
Setters and getters returns 'static' to make model easily extendable (just my over-caution). Can 'static' instead of 'self' impact on phalcon ORM? I got no any type errors. Changed it, the same error :smiling_face_with_tear:

**Reply by borisdelev** [reply]
I`ve just asking about "static"... anyway... Hmm... can u try something very simple in controller (for example):

```php
$test = new \Modules\Upload\Images\Image();
$test->is_vector = false;
$test->db_file_id = ''; // Enter some id from db_file table
$test->create();

var_dump($test->getMessages());
```

_Just want to debug from the base stuff. Is something little we didnt see_

**Reply by radynsade** [reply]
Tried this, got the same error

    $dbFile = new DbFile();
    $dbFile->link = '/test/link.png';
    $dbFile->size = 12345;
    $dbFile->mimeType = 'image/png';
    $dbFile->create();
    $image = new Image();
    $image->dbFileId = $dbFile->id;
    $image->dbFile = $dbFile;
    $image->isVector = false;
    $image->create();

var_dump:

    array(1) {
      [0]=>
      object(Phalcon\Messages\Message)#86 (5) {
        ["code":protected]=>
        int(0)
        ["field":protected]=>
        string(10) "db_file_id"
        ["message":protected]=>
        string(62) "Value of field "db_file_id" does not exist on referenced table"
        ["type":protected]=>
        string(19) "ConstraintViolation"
        ["metaData":protected]=>
        array(0) {
        }
      }
    }
    null

**Reply by borisdelev** [reply]
can u try like this:
```php
$dbFile = new DbFile();
$dbFile->link = '/test/link.png';
$dbFile->size = 12345;
$dbFile->mimeType = 'image/png';
$dbFile->create();
$image = new Image();
$image->db_file_id = $dbFile->id;
$image->is_vector = false;
$image->create();

var_dump($dbFile->getMessages());
var_dump($image->getMessages());
```

**Reply by radynsade** [reply]
I found where was a problem. In relationships declarations I specified column names instead of property names like $this->belongsTo('db_file_id... instead of $this->belongsTo('dbFileId... . After I changed it to property names, all works.

**Comment by radynsade** [comment, marked as answer]
I found where was a problem. In relationships declarations I specified column names instead of property names like $this->belongsTo('db_file_id... instead of $this->belongsTo('dbFileId... . After I changed it to property names, all works.
